cardano-api/src/Cardano/Api/Typed.hs

```diff
@@ -0,0 +1,1142 @@
+-- Keys key keys!
+--
+
+
+
+-- $keys
+-- Cardano has lots of cryptographic keys used for lots of different purposes.
+-- Some keys have different representations, but most are just using keys in
+-- different roles.
+--
+-- To allow for the different representations and to avoid mistakes we
+-- distinguish the key /role/. These are type level distinctions, so each of
+-- these roles is a type level tag.
+--
+
+-- | An interface for cryptographic keys used for signatures with a 'SigningKey'
+-- and a 'VerificationKey' key.
+--
+-- This interface does not provide actual signing or verifying functions since
+-- this API is concerned with the management of keys: generating and
+-- serialising.
+--
+class Key keyrole where
+
+    -- | The type of cryptographic verification key, for each key role.
+    data VerificationKey keyrole :: *
+
+    -- | The type of cryptographic signing key, for each key role.
+    data SigningKey keyrole :: *
+
+    -- | Get the corresponding verification key from a signing key.
+    getVerificationKey :: SigningKey keyrole -> VerificationKey keyrole
+
+    -- | Generate a 'SigningKey' deterministically, given a 'Crypto.Seed'. The
+    -- required size of the seed is given by 'deterministicSigningKeySeedSize'.
+    --
+    deterministicSigningKey :: AsType keyrole -> Crypto.Seed -> SigningKey keyrole
+    deterministicSigningKeySeedSize :: AsType keyrole -> Word
+
+--    verificationKeyHash :: VerificationKey keyrole -> Hash keyrole
+
+
+-- | Generate a 'SigningKey' using a seed from operating system entropy.
+--
+generateSigningKey :: Key keyrole => AsType keyrole -> IO (SigningKey keyrole)
+generateSigningKey keytype = do
+    seed <- Crypto.readSeedFromSystemEntropy seedSize
+    return $! deterministicSigningKey keytype seed
+  where
+    seedSize = deterministicSigningKeySeedSize keytype
+
+
+-- | Some key roles share the same representation and it is sometimes
+-- legitimate to change the role of a key.
+--
+class CastKeyRole keyroleA keyroleB where
+
+    -- | Change the role of a 'VerificationKey', if the representation permits.
+    castVerificationKey :: VerificationKey keyroleA -> VerificationKey keyroleB
+
+    -- | Change the role of a 'SigningKey', if the representation permits.
+    castSigningKey :: SigningKey keyroleA -> SigningKey keyroleB
+
+
+data family Hash keyrole :: *
+
+class CastHash keyroleA keyroleB where
+
+    castHash :: Hash keyroleA -> Hash keyroleB
+
+
+-- ----------------------------------------------------------------------------
+-- Addresses
+--
+
+data Address era where
+
+     -- | Byron addresses are valid in both the Byron and Shelley era.
+     --
+     ByronAddress
+       :: Byron.Address
+       -> Address era
+
+     -- | Shelley addresses are only valid in the Shelley era.
+     --
+     ShelleyAddress
+       :: PaymentCredential
+       -> StakeCredentialReference
+       -> NetworkId
+       -> Address Shelley
+
+data StakeAddress where
```

Comment:
This doesn't need to be a GADT. Are we leaving it as a GADT to allow for future flexibility?

Reply:
Strictly speaking it's not a GADT, it is just using GADT syntax. I just did it that way for consistency, it doesn't make any difference.